kernel/src/hil/uart.rs

```diff
@@ -44,35 +57,223 @@ pub enum Error {
     /// Read or write was aborted early
     Aborted,
 
-    /// No error occurred and the command completed successfully
-    CommandComplete,
 }
 
-pub trait UART {
-    /// Set the client for this UART peripheral. The client will be
-    /// called when events finish.
-    fn set_client(&self, client: &'static Client);
+pub trait Uart<'a>: Configure + Transmit<'a> + Receive<'a> {}
+pub trait UartData<'a>: Transmit<'a> + Receive<'a> {}
```

Comment:
This name has the unfortunate property that I read it where `Uart` is the adjective, and `Data` is the noun, and then it's not clear why a capsule needs some uart specific data (or what that even is). Adding `Bus` would make it clear, but that ruins the symmetry...

Reply:
Happy to change the name -- Could be ReceiveTransmit, and Uart could become FullUart?